### ISSUE TITLE
[$500] Tag - Child tags are not indented on list

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#31753]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.2-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

**Issue found when executing PR** https://github.com/Expensify/App/pull/31494
## Action Performed:
Precondition: enable and add parent and child tags as an admin on OD
1. Log in to ND as an employee
2. Navigate to workspace chat
3. Tap the "+" > Request money
4. Enter the amount
5. On the confirmation screen, tap the "Show more" button
6. Select "Tags"

## Expected Result:
Child tags are indented on the list

## Actual Result:
Child tags are not indented. It is a different behavior with categories: child categories are indented

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

![Bug6287714_1700685033775!photo_2023-11-22_14-24-49](https://github.com/Expensify/App/assets/78819774/e6c29a2d-87e3-4508-97c2-4b35890d12d0)

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01f5ae8cc738a492b2</li>
        <li>Upwork Job ID: 1727436127562772480</li>
        <li>Last Price Increase: 2023-11-22</li>
    </ul>
</details>

**Comment by wlegolas** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Tag - Child tags are not indented on list

### What is the root cause of that problem?

The method `getTagsOptions` is used to get the Tags to be displayed, this method does not create a tree structure when the tag name has the format `[parent]:[child]`. 

https://github.com/Expensify/App/blob/a5e4fb3f975e4e2a769ccef01b4b4fec4172bef6/src/libs/OptionsListUtils.js#L912-L925

To generate the Tag options is used the sequence of methods from the `OptionsListUtils`: `getFilteredOptions` -> `getOptions` -> `getTagListSections` -> `getTagsOptions`


### What changes do you think we should make in order to solve the problem?

We should change the `getTagsOptions` method to call the [getCategoryOptionTree](https://github.com/Expensify/App/blob/main/src/libs/OptionsListUtils.js#L756) method with the tags to create the tree structure when the tag name has the format `[parent]:[child]`.

For example,

```js
/**
 * Transforms the provided tags into objects with a specific structure.
 *
 * @param {Object[]} tags - an initial tag array
 * @param {Boolean} tags[].enabled - a flag to enable/disable option in a list
 * @param {String} tags[].name - a name of an option
 * @returns {Array<Object>}
 */
function getTagsOptions(tags) {
    const normalizedTags = _.map(tags, (tag) => {
        // This is to remove unnecessary escaping backslash in tag name sent from backend.
        const tagName = tag.name && tag.name.replace(/\\{1,2}:/g, ':');

        return {
            ...tag,
            name: tagName,
        };
    });

    return getCategoryOptionTree(normalizedTags);
}
```

### What alternative solutions did you explore? (Optional)

N/A

**Comment by MitchExpensify** [comment]
Having some trouble reproducing this, do you have a video of your reproduction @lanitochka17 ?

**Comment by c3024** [comment]
I used the mocking in [this comment](https://github.com/Expensify/App/pull/28151#issuecomment-1736543304) to get tags. However, I am unable to get multitags with the [steps here](https://community.expensify.com/discussion/5756/how-to-set-up-and-manage-multi-level-tagging/).

Could you share the csv file you uploaded to get multitags? @lanitochka17  @wlegolas

**Comment by lanitochka17** [comment]
@c3024 Hello
Multitags were entered manually with PR https://github.com/Expensify/App/pull/31494

**Comment by wlegolas** [comment]
> I used the mocking in [this comment](https://github.com/Expensify/App/pull/28151#issuecomment-1736543304) to get tags. However, I am unable to get multitags with the [steps here](https://community.expensify.com/discussion/5756/how-to-set-up-and-manage-multi-level-tagging/).
> 
> Could you share the csv file you uploaded to get multitags? @lanitochka17 @wlegolas

Hi @c3024 to help I created the code below to set the Tags for a specific Policy. You can get a Policy ID and execute the code below.

```js
Onyx.set('policyTags_C0C6E22CBEB45222', {
  Tag: {
    name: 'Custom',
    required: false,
    tags: {
        Movies: {
            enabled: true,
            name: 'Movies'
        },
        'Movies: Avengers: Endgame': {
            enabled: true,
            name: 'Movies: Avengers: Endgame',
            unencodedName: 'Movies: Avengers: Endgame'
        },
        Places: {
            enabled: true,
            name: 'Places'
        },
    }
  }
})
```

**Comment by MitchExpensify** [comment]
I am still unable to reproduce this, how are you all seeing Custom, all I see is Categories after clicking "Show more":

**Comment by wlegolas** [comment]
> I am still unable to reproduce this, how are you all seeing Custom, all I see is Categories after clicking "Show more":

Hi @MitchExpensify, to show the component to select the Tags, the condition below must be `true`:
https://github.com/Expensify/App/blob/679d770c439bc26e4d8877c3f556fb26d33fa36e/src/components/MoneyRequestConfirmationList.js#L242

Maybe the problem is opening the `MoneyRequestConfirmationList` without Tags for the related policy.

In the video below, I show how you can set the Tags for a specific policy. In this video, I got the Policy ID before starting the video. 

The first time I showed the confirm page without the Tags, and the next time I manually created the Tags for the specific Policy (using the script in this comment https://github.com/Expensify/App/issues/31753#issuecomment-1828733882), and the Tags field is shown on the confirm page.

**Note**: in this video, you can see the Tags are not shown in the format `[parent]:[child]`, I've been using a branch without the changes to show in the correct format.

I hope this video helps you to validate this issue.

https://github.com/Expensify/App/assets/698363/7ece96a2-2abe-4412-8804-9cb705aea1ae

**Comment by c3024** [comment]
@wlegolas Your RCA is correct and the proposal looks good. 

But when we search for a child in Category Page there is no tree structure shown in the results but in Tag Page there is. Can we make this consistent?

<img width="398" alt="Screenshot 2023-11-29 at 10 48 58 AM" src="https://github.com/Expensify/App/assets/102477862/8499ba95-329f-42ef-9fde-2fe4669399cc">


https://github.com/Expensify/App/assets/102477862/8d439675-57bb-44b8-9079-a798c5c284c0

**Comment by c3024** [comment]
I think the search results indentation in Tag page is the correct way and changing the search results indentation in Category page is not within the scope of this issue.

So, @wlegolas [proposal here](https://github.com/Expensify/App/issues/31753#issuecomment-1823783910) looks good to me.

🎀 👀 🎀 C+ reviewed

**Comment by neil-marcellini** [comment]
> I think the search results indentation in Tag page is the correct way and changing the search results indentation in Category page is not within the scope of this issue.
> 
> So, @wlegolas [proposal here](https://github.com/Expensify/App/issues/31753#issuecomment-1823783910) looks good to me.
> 
> 🎀 👀 🎀 C+ reviewed

I agree. The only thing is I think we should rename the function `getCategoryOptionTree` to something more general like `getIndentedOptionTree`.

**Comment by wlegolas** [comment]
Hello there.

Just to keep you up to date, I'm facing some problems running Android in my environment, maybe tomorrow I can send the PR here.

**Comment by MitchExpensify** [comment]
How is progress here?

**Comment by c3024** [comment]
PR merged and deployed to staging.

**Comment by MitchExpensify** [comment]
nice one!

**Comment by wlegolas** [comment]
> Compensation has been sent for your work all the same! Thanks

Hi, @MitchExpensify, thank you for sharing this context.

Thank you very much for the payment, even after it was completed and you saw that it was not a bug, this shows the great respect that Expensify has with the collaborators.
